Add Binary Messaging Interface Support into Android Extension

huningxin opened the pull request:
BUG=XWALK-5019
BUG=XWALK-4615

huningxin commented:
@halton @DonnaWuDongxia , PTAL. Thanks!

rakuco commented on extensions/common/android/xwalk_extension_android.cc:
```diff
@@ -162,21 +178,33 @@ XWalkExtensionAndroidInstance::~XWalkExtensionAndroidInstance() {
+    ScopedJavaLocalRef<jstring> buffer(env, env->NewStringUTF(value.c_str()));
+    Java_XWalkExtensionAndroid_onMessage(
+        env, obj.obj(), getID(), buffer.obj());
+    return;
```
All these new `return` statements aren't needed. You could rewrite this block like this:

```c++
if (GetAsString(...)) {
} else if (GetAsBinary(...) {
} else {
  NOTREACHED() << "Failed to decode message as either string or binary blob";
}
```

You could even replace the `NOTREACHED` with something more fatal if necessary; for example, is it possible that even decoding a binary blob fails?

huningxin replied:
It sounds right to me. I will update.

rakuco commented:
I think you should reference the JIRA tickets from the individual commits in the pull request message as well, otherwise there will be no notification of this PR in JIRA.

huningxin commented:
Update the PR to address comments from @rakuco . Please take another look. Thanks!

Ping @halton for Android and extension part. @DonnaWuDongxia for extension. @lincsoon for Android part. Thanks!

huningxin commented:
> I think you should reference the JIRA tickets from the individual commits in the pull request message as well, otherwise there will be no notification of this PR in JIRA.

Thanks for the suggestion. I put them into the comments. But it seems still no notifications in JIRA. Anyway, I update the the JIRA manually.

rakuco commented:
The comment about a pull request referencing a certain JIRA issue is only added when the pull request is created, which is why nothing happened now.
However, if/when the PR is merged, the related JIRA issues will be automatically closed.

rakuco commented on extensions/common/android/xwalk_extension_android.cc:
```diff
@@ -162,21 +178,32 @@ XWalkExtensionAndroidInstance::~XWalkExtensionAndroidInstance() {
+    NOTREACHED() << "Failed to decode message as either string or binary blob";
+  }
+
+  return;
```
This `return` here is also unnecessary, you're about to leave the function anyway.

huningxin replied:
yup, will delete it.

huningxin commented:
@rakuco , addressed your comment in the latest PR. Please take another look. Thanks!

Ping @halton for Android and extension part. @DonnaWuDongxia for extension. @lincsoon for Android part. Thanks!

DonnaWuDongxia commented:
So this PR is just enabling BMI for Android. BMI for c/c++ level is now landed in extension-sdk repo, not in crosswalk, right?

DonnaWuDongxia commented:
With this PR, new interfaces "postBinaryMessage", "onBinaryMessage" are intruduced to extension implementation side, and JS stub never aware any difference, and accurally, "postBinaryMessage" and "postMessage" are used the same channel before Android level. So, can we just make the "binary message" a special "message" by handling them in the same interface with difference branches, or change the parameter of postMesasge/onMessage to contain both string and binary? So that, we can also unaware the difference on the extension implementation side.

halton commented:
No, the C++ BMI support has been added with PR #3155

huningxin commented:
> So this PR is just enabling BMI for Android. BMI for c/c++ level is now landed in extension-sdk repo, not in crosswalk, right?

huningxin commented:
> So, can we just make the "binary message" a special "message" by handling them in the same interface with difference branches, or change the parameter of postMesasge/onMessage to contain both string and binary?

`postMessage` and `postBinaryMessage` support primitive message types (test data in `String` and binary data in `byte[]`). Embedders and tools (as JsStub) are free to build upper abstraction on top of them. For example, as you described, upper tools are able to build unified handling for both text and binary messages. This PR focuses on the low level binary message type support.

Again, this is the Android part implementation of https://github.com/crosswalk-project/crosswalk/pull/3155 and the design discussion was on:

https://lists.crosswalk-project.org/pipermail/crosswalk-dev/2015-July/002932.html

https://lists.crosswalk-project.org/pipermail/crosswalk-dev/2015-July/002931.html

huningxin commented:
Update the PR to resolve the conflicts and address @DonnaWuDongxia 's comment.

Please take another look. Thanks!

DonnaWuDongxia commented:
@huningxin I suppose the interface postBinaryMessage is not avaliable to jsStub context, right?

huningxin commented:
I understand current jsStub is built on top of postMessage. If jsStub wants to support binary message passing without any text-binary encoding, it needs to add postBinaryMessage support.

DonnaWuDongxia commented:
Generally, the PR is LGTM. I'm trying to understand the differences brought by this PR. In the previous comment, I mentioned "jsStub", this term accually means the js part of the extension implementation. So, as a extension developer, the interfaces are not changed in the js side, and two new interfaces in the native(c/c++/java) part, right?

DonnaWuDongxia commented:
In another word, existing "postMessage" interface in jsStub can be used to send binary data to the native part, no need to add another interface dedicate to post binary data from js to native. Right?

huningxin commented:
Right. As shown in the sample code, JavaScript code still uses `extension.postMessage` API but is able to post binary data wrapped in an ArrayBuffer. Same as JS message handler.

DonnaWuDongxia commented:
Thanks! LGTM!

huningxin commented:
@halton @rakuco  do you have any more comments?

halton commented on xwalk_android_tests.gypi:
```diff
@@ -148,6 +148,7 @@
           '<(PRODUCT_DIR)/xwalk_xwview_test/assets/console_message.html',
           '<(PRODUCT_DIR)/xwalk_xwview_test/assets/echo_java.html',
           '<(PRODUCT_DIR)/xwalk_xwview_test/assets/echo_sync_java.html',
+          '<(PRODUCT_DIR)/xwalk_xwview_test/assets/echo_binary_java.html',
```
nit, order

halton commented on xwalk_android_tests.gypi:
```diff
@@ -173,6 +174,7 @@
             'test/android/data/console_message.html',
             'test/android/data/echo_java.html',
             'test/android/data/echo_sync_java.html',
+            'test/android/data/echo_binary_java.html',
```
ditto

halton commented on xwalk_android_tests.gypi:
```diff
@@ -451,6 +453,7 @@
           '<(PRODUCT_DIR)/runtime_client_test/assets/echo.html',
           '<(PRODUCT_DIR)/runtime_client_test/assets/echo_java.html',
           '<(PRODUCT_DIR)/runtime_client_test/assets/echo_sync_java.html',
+          '<(PRODUCT_DIR)/runtime_client_test/assets/echo_binary_java.html',
```
order

halton commented on xwalk_android_tests.gypi:
```diff
@@ -468,6 +471,7 @@
             'test/android/data/displayAvailableTest.html',
             'test/android/data/echo_java.html',
             'test/android/data/echo_sync_java.html',
+            'test/android/data/echo_binary_java.html',
```
order

halton commented:
LGTM, please fix the order issue so that I can merge.

huningxin commented:
Hi @halton , updated the PR to address your comments. Please take another look. Thanks!

halton commented on xwalk_android_tests.gypi:
```diff
@@ -147,6 +147,7 @@
           '<(PRODUCT_DIR)/xwalk_xwview_test/assets/create_window_2.html',
           '<(PRODUCT_DIR)/xwalk_xwview_test/assets/console_message.html',
           '<(PRODUCT_DIR)/xwalk_xwview_test/assets/echo_java.html',
+          '<(PRODUCT_DIR)/xwalk_xwview_test/assets/echo_binary_java.html',
```
@huningxin One trial comment on  the order. echo_binary_java should before echo_java. Ditto for rest.

huningxin replied:
Sorry for this. will fix them.

halton commented:
LGTM, will merge after trybot green.

huningxin commented:
Thanks for review, @halton ! Now all bots are green.

rakuco commented:
There are problems with the new test: https://build.crosswalk-project.org/builders/Crosswalk%20Tests%20%28Android%20x86%29/builds/1614/steps/xwalk_core_test/logs/stdio and https://build.crosswalk-project.org/builders/Crosswalk%20Tests%20%28Android%20x86%29/builds/1614/steps/xwalk_core_internal_test/logs/stdio

Is anyone working on fixing them? Otherwise we'll have to revert the patch.

huningxin commented:
@rakuco , i am looking into it.

huningxin commented:
@rakuco , i created a PR to fix the testing failures: https://github.com/crosswalk-project/crosswalk/pull/3275
